x-pack/elastic-agent/pkg/agent/application/fleet_gateway.go

```diff
@@ -219,6 +237,18 @@ func (f *fleetGateway) execute(ctx context.Context) (*fleetapi.CheckinResponse,
 	return resp, nil
 }
 
+func (f *fleetGateway) shouldUnroll() bool {
+	return f.unauthCounter >= maxUnauthCounter
+}
+
+func isUnauth(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	return strings.Contains(err.Error(), fleetapi.ErrInvalidAPIKey.Error())
```

Comment:
Is it not possible to unwrap the error and check for the specific error type?

Would prefer that over string matching.

Reply:
i spent some time trying to achieve that but with no luck, i will try to spend some time with it again